ctapipe/core/qualityquery.py

```diff
@@ -25,38 +31,45 @@ class QualityQuery(Component):
     """
 
     quality_criteria = List(
-        Tuple(Unicode(), Unicode()),
-        help=(
-            "list of tuples of ('<description', 'expression string') to accept "
-            "(select) a given data value.  E.g. ``[('mycut', 'x > 3'),]``. "
-            "You may use ``numpy`` as ``np`` and ``astropy.units`` as ``u``,"
-            " but no other modules."
-        ),
+        Tuple(Unicode(), TelescopeParameter(Unicode())),
+        help="List of tuples of ('query name', TelescopeParameter or expression) to accept"
+        " (select) a given data value. Example: ``[('positive', 'x > 0')]`` or"
+        " ``[('high_intensity', [('type', '*', 'hillas_intensity > 50'), ('type', 'SST*', 'hillas_intensity > 100')])]``"
+        " You may use ``numpy`` as ``np`` and ``astropy.units`` as ``u``,"
+        " but no other modules.",
     ).tag(config=True)
 
-    def __init__(self, config=None, parent=None, **kwargs):
-        super().__init__(config=config, parent=parent, **kwargs)
+    def __init__(self, subarray, config=None, parent=None, **kwargs):
+        super().__init__(subarray=subarray, config=config, parent=parent, **kwargs)
 
-        # add a selection to count all entries and make it the first one
-        self.criteria_names = [n for (n, _) in self.quality_criteria]
-        self.expressions = [e for (_, e) in self.quality_criteria]
+        self._compile()
 
-        self.engine = ExpressionEngine(self.quality_criteria)
-        for _, expr in self.quality_criteria:
-            if "lambda" in expr:
-                raise ValueError(
-                    "As of ctapipe 0.16, do not give lambda expressions"
-                    " to QualityQuery. Directly give the expression."
-                    " E.g. instead of `lambda p: p.hillas.width.value > 0`"
-                    " use `parameters.hillas.width.value > 0`"
-                )
+    def _compile(self):
+        self.engines = []
+        self.criteria_names = []
+
+        for name, criteria_list in self.quality_criteria:
+
+            if not isinstance(criteria_list, (tuple, list, TelescopePatternList)):
+                criteria_list = TelescopePatternList([criteria_list])
+
+            for i, crit in enumerate(criteria_list):
+                command, arg, expr = crit
+                par = (command, arg, ExpressionEngine(((name, expr),)))
+                criteria_list[i] = par
+
+            self.criteria_names.append(name)
+
+            lookup = TelescopeParameterLookup(criteria_list)
+            lookup.attach_subarray(self.subarray)
+            self.engines.append((name, lookup))
 
         # arrays for recording overall statistics, add one for total count
-        n = len(self.quality_criteria) + 1
+        n = len(self.criteria_names) + 1
         self._counts = np.zeros(n, dtype=np.int64)
         self._cumulative_counts = np.zeros(n, dtype=np.int64)
 
-    def to_table(self, functions=False):
+    def to_table(self, functions=False) -> Table:
```

Comment:
kwarg now unused?

Reply:
The processtool uses it: https://github.com/cta-observatory/ctapipe/blob/c1f7ba0994e485d983d5fd861d55563682c0c591/ctapipe/tools/process.py#L283

We need to decide whether (and how) to support this.  I left this here as unused kwarg to not change to ProcessTool

Reply:
I meant inside the function body.

So I'd say that this info is available from the configuration, which is stored in provenance. Alternatively, it could be stored as metadata on the table, not in the table itself?

Reply:
I think config and provenance is enough